llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1750,6 +1752,88 @@ void lowerGlobalsToVector(Module &M) {
 
 } // namespace
 
+static void checkSLMInit(Module &M) {
+  SmallPtrSet<const Function *, 8u> Callers;
+  bool Kernel_Has_slm_init = false;
+  bool Kernel_Has_local_accessor = false;
+
+  for (auto &F : M) {
+    if (!isSlmInit(F)) {
+      if (Kernel_Has_local_accessor) {
+        continue;
+      }
+      if (F.getName().starts_with(SPIRV_LOCAL_ACCESSOR_PREF)) {
+        Kernel_Has_local_accessor = true;
+        continue;
+      }
+      unsigned Idx = 0;
+      for (const Argument &Arg : F.args()) {
+        if (Arg.getType()->isPointerTy()) {
+          auto *KernelArgAccPtrs = F.getMetadata("kernel_arg_accessor_ptr");
```

Comment:
Can we send an email to the CFE/SYCL language team as a non-blocking follow-up to see if they have any ideas on improving this check by doing it somewhere else (Sema?). Relying on the metadata should work but it seems a bit indirect and possibly flaky, but I don't know how to do any better today.